In [91]:
# Yihao Zhong Larry, yz7654, date: 02/23/2024 - onwards
# 

## Problem 1 -  Softmax Activation Function

### 1.1. logarithmic derivative


Given that for each real-valued outputs $v_1$, $v_2$, ..., $v_k$, $o_i = \frac{exp(v_i)}{\sum^k_{j=1}{exp(v_j)}}$

For the case where $i=j$, we consider the derivative of $o_i$ with respect to $v_i$.

Take logarithm: $$ln(o_i) = ln(\frac{exp(v_i)}{\sum^k_{j=1}{exp(v_j)}}) = v_i-ln(\sum^k_{j=1}{exp(v_j)})$$

Taking the derivative of both sides with respect to $v_i$: 
$$\frac{d}{dv_i} ln(o_i) = 1 - \frac{1}{\sum^k_{j=1}{exp(v_j)}}e^{v_i}$$

Since $o_i = \frac{exp(v_i)}{\sum^k_{j=1}{exp(v_j)}}$, we have $$ \frac{d}{dv_i} ln(o_i) = 1-o_i $$

To find $\frac{\partial{o_i}}{\partial{v_j}}$, we know that $\frac{\partial}{\partial{x_j}}ln(f(x)) = \frac{1}{f(x)}\frac{\partial{f(x)}}{\partial{x_j}}$ by logarithmic derivative.

Therefore, $$\frac{\partial{o_i}}{\partial{v_j}} = o_i(1-o_i)$$


For the case $i \neq j$, we have start similarly by applying the logarithm to $i$, but take the derivative with respect to $v_j$.

Take logarithm: $$ln(o_i) = ln(\frac{exp(v_i)}{\sum^k_{j=1}{exp(v_j)}})$$

Taking the derivative with respect to $v_j$: $$\frac{d}{dv_j} ln(o_i) =-\frac{e^{v_j}}{\sum^k_{j=1}{exp(v_j)}}$$ 

Simply this by definition of $v_j$ we have: $$\frac{d}{dv_j} ln(o_i) = -o_j$$

To find $\frac{\partial{o_i}}{\partial{v_j}}$, we know that $\frac{\partial}{\partial{x_j}}ln(f(x)) = \frac{1}{f(x)}\frac{\partial{f(x)}}{\partial{x_j}}$ by logarithmic derivative.

Therefore, $$\frac{\partial{o_i}}{\partial{v_j}} = -o_io_j$$

### 1.2. 

## Problem 2 -  Neural Network Training and Backpropagation

In [92]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from scipy.io import loadmat
%matplotlib inline

data = loadmat('data/ex3data1.mat')
X = data['X']
y = data['y']

print(X.shape, y.shape)

from sklearn.preprocessing import OneHotEncoder
encoder = OneHotEncoder(sparse=False)
y_onehot = encoder.fit_transform(y)
print(y_onehot.shape)

(5000, 400) (5000, 1)
(5000, 10)


/Users/zhongyihao/anaconda3/lib/python3.10/site-packages/sklearn/preprocessing/_encoders.py:828: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(


### 2.1.

In [93]:
def scaled_sigmoid(z):
    return 1 / (1 + np.exp(-2 * z))


### 2.2. forward_propagate()

In [94]:
def forward_propagate(X, theta1, theta2, theta3):
    m = X.shape[0]
    
    # Input layer activations (adding bias unit)
    a1 = np.insert(X, 0, values=np.ones(m), axis=1)

    z2 = a1 * (theta1.T)
    a2 = np.insert(scaled_sigmoid(z2), 0, values=np.ones(m), axis=1)
    
    z3 = a2 * (theta2.T)
    a3 = np.insert(scaled_sigmoid(z3), 0, values=np.ones(m), axis=1)
    
    z4 = a3 * (theta3.T)
    h = scaled_sigmoid(z4)
    
    return a1, z2, a2, z3, a3, z4, h


### 2.3. cost()

In [95]:
def cost(params, input_size, hidden_size1, hidden_size2, num_labels, X, y, learning_rate):
    m = X.shape[0]
    X = np.matrix(X)
    y = np.matrix(y)
    
    end_theta1 = hidden_size1 * (input_size + 1)
    end_theta2 = end_theta1 + hidden_size2 * (hidden_size1 + 1)
    theta1 = np.reshape(params[:end_theta1], (hidden_size1, input_size + 1))
    theta2 = np.reshape(params[end_theta1:end_theta2], (hidden_size2, hidden_size1 + 1))
    theta3 = np.reshape(params[end_theta2:], (num_labels, hidden_size2 + 1))
    
    # just need the output of the last layer
    _, _, _, _, _, _, h = forward_propagate(X, theta1, theta2, theta3)
    
    # Compute the cost
    J = 0
    for i in range(m):
        first_term = np.multiply(-y[i, :], np.log(h[i, :]))
        second_term = np.multiply((1 - y[i, :]), np.log(1 - h[i, :]))
        J += np.sum(first_term - second_term)
    J = J / m

    return J

def cost_with_regulazation(params, input_size, hidden_size1, hidden_size2, num_labels, X, y, learning_rate):
    m = X.shape[0]
    X = np.matrix(X)
    y = np.matrix(y)
    
    end_theta1 = hidden_size1 * (input_size + 1)
    end_theta2 = end_theta1 + hidden_size2 * (hidden_size1 + 1)
    theta1 = np.reshape(params[:end_theta1], (hidden_size1, input_size + 1))
    theta2 = np.reshape(params[end_theta1:end_theta2], (hidden_size2, hidden_size1 + 1))
    theta3 = np.reshape(params[end_theta2:], (num_labels, hidden_size2 + 1))
    
    # just need the output of the last layer
    _, _, _, _, _, _, h = forward_propagate(X, theta1, theta2, theta3)
    
    # Compute the cost
    J = 0
    for i in range(m):
        first_term = np.multiply(-y[i, :], np.log(h[i, :]))
        second_term = np.multiply((1 - y[i, :]), np.log(1 - h[i, :]))
        J += np.sum(first_term - second_term)
    J = J / m

    J += (float(learning_rate) / (2 * m)) \
            * (np.sum(np.square(theta1[:, 1:])) \
                + np.sum(np.square(theta2[:, 1:])) + np.sum(np.square(theta3[:, 1:])))
    return J



In [96]:
# Initial setup
input_size = 400  # Size of input layer
hidden_size1 = 20  # Size of first hidden layer
hidden_size2 = 20  # Size of second hidden layer
num_labels = 10  # Size of output layer
learning_rate = 1

m = X.shape[0]
X = np.matrix(X)
y = np.matrix(y)

# Randomly initialize parameters for the full network's parameters, adjusted for 3 layers
params_size = (input_size + 1) * hidden_size1 + (hidden_size1 + 1) * hidden_size2 + (hidden_size2 + 1) * num_labels
params = (np.random.random(size=params_size) - 0.5) * 0.25

# Unravel the parameter array into parameter matrices for each layer
end_theta1 = hidden_size1 * (input_size + 1)
end_theta2 = end_theta1 + hidden_size2 * (hidden_size1 + 1)

theta1 = np.matrix(np.reshape(params[:end_theta1], (hidden_size1, input_size + 1)))
theta2 = np.matrix(np.reshape(params[end_theta1:end_theta2], (hidden_size2, hidden_size1 + 1)))
theta3 = np.matrix(np.reshape(params[end_theta2:], (num_labels, hidden_size2 + 1)))

print(f"Theta1 shape: {theta1.shape}")  
print(f"Theta2 shape: {theta2.shape}")  
print(f"Theta3 shape: {theta3.shape}")  

Theta1 shape: (25, 401)
Theta2 shape: (25, 26)
Theta3 shape: (10, 26)


In [97]:
print("cost without regularization: ", 
      cost(params, input_size, hidden_size1, hidden_size2, num_labels, X, y_onehot, learning_rate))
print("cost with regularization: ", 
      cost_with_regulazation(params, input_size, hidden_size1, hidden_size2, num_labels, X, y_onehot, learning_rate))
a1, z2, a2, z3, a3, z4, h = forward_propagate(X, theta1, theta2, theta3)
print(f"a1: {a1.shape}, z2: {z2.shape}, a2: {a2.shape}, z3: {z3.shape}, a3: {a3.shape}, z4: {z4.shape}, h: {h.shape}")

cost without regularization:  7.447201218857694
cost with regularization:  7.452858423785688
a1: (5000, 401), z2: (5000, 25), a2: (5000, 26), z3: (5000, 25), a3: (5000, 26), z4: (5000, 10), h: (5000, 10)


### 2.4. sigmoid gradient



In [98]:

def scaled_sigmoid_gradient(z):
    return np.multiply(scaled_sigmoid(z), (1 - scaled_sigmoid(z)))

### 2.5. backprop()

In [99]:
def get_theta_shapes(input_size, hidden_size1, hidden_size2, num_labels):
    """
    Returns the shapes of theta matrices for each layer in the network.
    """
    theta1_shape = (hidden_size1, input_size + 1)
    theta2_shape = (hidden_size2, hidden_size1 + 1)
    theta3_shape = (num_labels, hidden_size2 + 1)
    return theta1_shape, theta2_shape, theta3_shape

def reshape_params(params, shapes):
    """
    Reshapes the flattened parameter array into individual theta matrices.
    """
    theta1_shape, theta2_shape, theta3_shape = shapes
    size1 = theta1_shape[0] * theta1_shape[1]
    size2 = size1 + theta2_shape[0] * theta2_shape[1]
    
    theta1 = params[:size1].reshape(theta1_shape)
    theta2 = params[size1:size2].reshape(theta2_shape)
    theta3 = params[size2:].reshape(theta3_shape)
    
    return theta1, theta2, theta3


In [100]:
def backprop(params, input_size, hidden_size1, hidden_size2, num_labels, X, y, learning_rate):
    m = X.shape[0]
    X = np.matrix(X)
    y = np.matrix(y)
    
    theta_shapes = get_theta_shapes(input_size, hidden_size1, hidden_size2, num_labels)
    theta1, theta2, theta3 = reshape_params(params, theta_shapes)
    
    a1, z2, a2, z3, a3, z4, h = forward_propagate(X, theta1, theta2, theta3)
    
    # Gradients initialization
    delta1 = np.zeros(theta1.shape)
    delta2 = np.zeros(theta2.shape)
    delta3 = np.zeros(theta3.shape)

    # Compute cost
    J = cost_with_regulazation(params, input_size, hidden_size1, hidden_size2, num_labels, X, y, learning_rate)
    
    # Compute gradients (Backpropagation)
    for t in range(m):
        a1t = a1[t,:] 
        z2t = z2[t,:] 
        a2t = a2[t,:]  
        ht = h[t,:]  
        yt = y[t,:]
        
        d4 = ht - yt
        d3 = np.multiply((theta3.T * d4.T).T, scaled_sigmoid_gradient(np.insert(z3[t], 0, values=1)))
        d2 = np.multiply((theta2.T * d3[:,1:].T).T, scaled_sigmoid_gradient(np.insert(z2[t], 0, values=1)))
        
        delta3 += d4.T * a3[t]
        delta2 += d3[:,1:].T * a2t  
        delta1 += d2[:,1:].T * a1t  
    
    delta1 = delta1 / m
    delta2 = delta2 / m
    delta3 = delta3 / m
    
    # Unroll gradients
    grad = np.concatenate((np.ravel(delta1), np.ravel(delta2), np.ravel(delta3)))
    
    
    return J, grad

def backprop_with_reg(params, input_size, hidden_size1, hidden_size2, num_labels, X, y, learning_rate):
    m = X.shape[0]
    X = np.matrix(X)
    y = np.matrix(y)
    
    theta_shapes = get_theta_shapes(input_size, hidden_size1, hidden_size2, num_labels)
    theta1, theta2, theta3 = reshape_params(params, theta_shapes)
    
    a1, z2, a2, z3, a3, z4, h = forward_propagate(X, theta1, theta2, theta3)
    
    # Gradients initialization
    delta1 = np.zeros(theta1.shape)
    delta2 = np.zeros(theta2.shape)
    delta3 = np.zeros(theta3.shape)

    # Compute cost
    J = cost_with_regulazation(params, input_size, hidden_size1, hidden_size2, num_labels, X, y, learning_rate)
    
    # Compute gradients (Backpropagation)
    for t in range(m):
        a1t = a1[t,:] 
        z2t = z2[t,:] 
        a2t = a2[t,:]  
        ht = h[t,:]  
        yt = y[t,:]
        
        d4 = ht - yt
        d3 = np.multiply((theta3.T * d4.T).T, scaled_sigmoid_gradient(np.insert(z3[t], 0, values=1)))
        d2 = np.multiply((theta2.T * d3[:,1:].T).T, scaled_sigmoid_gradient(np.insert(z2[t], 0, values=1)))
        
        delta3 += d4.T * a3[t]
        delta2 += d3[:,1:].T * a2t  
        delta1 += d2[:,1:].T * a1t  
    
    delta1 = delta1 / m
    delta2 = delta2 / m
    delta3 = delta3 / m
    
    delta1[:, 1:] += (theta1[:, 1:] * learning_rate) / m
    delta2[:, 1:] += (theta2[:, 1:] * learning_rate) / m
    delta3[:, 1:] += (theta3[:, 1:] * learning_rate) / m

    # Unroll gradients
    grad = np.concatenate((np.ravel(delta1), np.ravel(delta2), np.ravel(delta3)))
    
    return J, grad


In [101]:
J, grad = backprop(params, input_size, hidden_size1, hidden_size2, num_labels, X, y_onehot, learning_rate)
print(f"Backpropagation cost without regularization: {J}, grad shape: {grad.shape}")

J_reg, grad_reg = backprop_with_reg(params, input_size, hidden_size1, hidden_size2, num_labels, X, y_onehot, learning_rate)
print(f"Backpropagation cost with regularization: {J_reg}, grad shape: {grad_reg.shape}")

Backpropagation cost without regularization: 7.452858423785688, grad shape: (10935,)


Backpropagation cost with regularization: 7.452858423785688, grad shape: (10935,)


### 2.6. & 2.7. Train

In [102]:
from scipy.optimize import minimize

options = {'maxiter': 250}


result = minimize(fun=backprop_with_reg, x0=params, 
                  args=(input_size, hidden_size1, hidden_size2, num_labels, X, y_onehot, learning_rate), 
                  method='TNC', jac=True, options=options)


params_optimized = result.x
cost_optimized = result.fun

print(result)
print(f"Optimized cost: {cost_optimized}")


/var/folders/v1/6k_h9wg90q56lxft6dxycw200000gn/T/ipykernel_26624/488540519.py:6: DeprecationWarning: 'maxiter' has been deprecated in favor of 'maxfun' and will be removed in SciPy 1.11.0.
  result = minimize(fun=backprop_with_reg, x0=params,


 message: Max. number of function evaluations reached
 success: False
  status: 3
     fun: 0.13757528041058906
       x: [-4.072e-01 -7.874e-04 ...  8.485e-01  9.292e-01]
     nit: 33
     jac: [ 2.072e-05 -1.575e-07 ...  3.179e-04  4.110e-04]
    nfev: 250
Optimized cost: 0.13757528041058906


In [103]:
end_theta1 = hidden_size1 * (input_size + 1)
end_theta2 = end_theta1 + hidden_size2 * (hidden_size1 + 1)


theta1 = np.matrix(np.reshape(result.x[:end_theta1], (hidden_size1, input_size + 1)))
theta2 = np.matrix(np.reshape(result.x[end_theta1:end_theta2], (hidden_size2, hidden_size1 + 1)))
theta3 = np.matrix(np.reshape(result.x[end_theta2:], (num_labels, hidden_size2 + 1)))


X = np.matrix(X)
# X = np.insert(X, 0, values=np.ones(X.shape[0]), axis=1)

a1, z2, a2, z3, a3, z4, h = forward_propagate(X, theta1, theta2, theta3)


y_pred = np.array(np.argmax(h, axis=1) + 1)

# y = np.squeeze(np.array(y))

correct = [1 if a == b else 0 for (a, b) in zip(y_pred, y)]
accuracy = (sum(correct) / float(len(correct)))
print('Accuracy = {0}%'.format(accuracy * 100))

Accuracy = 99.86%


### 2.8. 

We achieve an accuracy of 99.86% with a 3-layer neural network compared to 99.2% with a 2-layer network from the notebooks. Basically, we add an additional hidden layer increases the complexity of the model. This means the network has more parameters (weights and biases) to adjust during training, allowing it to capture more intricate patterns in the data. A deeper network can learn more complex function. By adding more layers, the network can create more complex representations of the input data.

## Problem 3 - Weight Initialization, Dead Neurons, Leaky ReLU

### 3.1. 


In [108]:

from keras.models import Sequential
from keras.layers import Conv2D, MaxPooling2D, Dense, Dropout, Flatten
from keras import backend as K

from matplotlib import pyplot as plt
from matplotlib import rcParamsDefault

## use of code from github.com/intoli/intoli-article-materials/blob/master/articles/neural-network-initialization/utils.py

def grid_axes_it(n_plots, n_cols=3, enumerate=False, fig=None):
    """
    Iterate through Axes objects on a grid with n_cols columns and as many
    rows as needed to accommodate n_plots many plots.

    Args:
        n_plots: Number of plots to plot onto figure.
        n_cols: Number of columns to divide the figure into.
        fig: Optional figure reference.

    Yields:
        n_plots many Axes objects on a grid.
    """
    n_rows = n_plots / n_cols + int(n_plots % n_cols > 0)

    if not fig:
        default_figsize = rcParamsDefault['figure.figsize']
        fig = plt.figure(figsize=(
            default_figsize[0] * n_cols,
            default_figsize[1] * n_rows
        ))

    for i in range(1, n_plots + 1):
        ax = plt.subplot(n_rows, n_cols, i)
        yield ax


def create_mlp_model(
    n_hidden_layers,
    dim_layer,
    input_shape,
    n_classes,
    kernel_initializer,
    bias_initializer,
    activation,
):
    """Create Multi-Layer Perceptron with given parameters."""
    model = Sequential()
    model.add(Dense(dim_layer, input_shape=input_shape, kernel_initializer=kernel_initializer,
                    bias_initializer=bias_initializer))
    for i in range(n_hidden_layers):
        model.add(Dense(dim_layer, activation=activation, kernel_initializer=kernel_initializer,
                        bias_initializer=bias_initializer))
    model.add(Dense(n_classes, activation='softmax', kernel_initializer=kernel_initializer,
                    bias_initializer=bias_initializer))
    return model


def create_cnn_model(input_shape, num_classes, kernel_initializer='glorot_uniform',
                     bias_initializer='zeros'):
    """Create CNN model similar to
       https://github.com/keras-team/keras/blob/master/examples/mnist_cnn.py."""
    model = Sequential()
    model.add(Conv2D(32, kernel_size=(3, 3),
                     activation='relu',
                     input_shape=input_shape,
                     kernel_initializer=kernel_initializer,
                     bias_initializer=bias_initializer))
    model.add(Conv2D(64, (3, 3), activation='relu',
                     kernel_initializer=kernel_initializer,
                     bias_initializer=bias_initializer))
    model.add(MaxPooling2D(pool_size=(2, 2)))
    model.add(Dropout(0.25))
    model.add(Flatten())
    model.add(Dense(128, activation='relu',
                    kernel_initializer=kernel_initializer,
                    bias_initializer=bias_initializer))
    model.add(Dropout(0.5))
    model.add(Dense(num_classes, activation='softmax',
                    kernel_initializer=kernel_initializer,
                    bias_initializer=bias_initializer))
    return model


def compile_model(model):
    model.compile(loss=keras.losses.categorical_crossentropy,
                  optimizer=keras.optimizers.RMSprop(),
                  metrics=['accuracy'])
    return model


def get_init_id(init):
    """
    Returns string ID summarizing initialization scheme and its parameters.

    Args:
        init: Instance of some initializer from keras.initializers.
    """
    try:
        init_name = str(init).split('.')[2].split(' ')[0]
    except:
        init_name = str(init).split(' ')[0].replace('.', '_')

    param_list = []
    config = init.get_config()
    for k, v in config.items():
        if k == 'seed':
            continue
        param_list.append('{k}-{v}'.format(k=k, v=v))
    init_params = '__'.join(param_list)

    return '|'.join([init_name, init_params])


def get_activations(model, x, mode=0.0):
    """Extract activations with given model and input vector x."""
    outputs = [layer.output for layer in model.layers]
    activations = K.function([model.input], outputs)
    output_elts = activations([x, mode])
    return output_elts


class LossHistory(keras.callbacks.Callback):
    """A custom keras callback for recording losses during network training."""

    def on_train_begin(self, logs={}):
        self.losses = []
        self.epoch_losses = []
        self.epoch_val_losses = []

    def on_batch_end(self, batch, logs={}):
        self.losses.append(logs.get('loss'))

    def on_epoch_end(self, epoch, logs={}):
        self.epoch_losses.append(logs.get('loss'))
        self.epoch_val_losses.append(logs.get('val_loss'))

Unexpected exception formatting exception. Falling back to standard exception


Traceback (most recent call last):
  File "/Users/zhongyihao/anaconda3/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 3505, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "/var/folders/v1/6k_h9wg90q56lxft6dxycw200000gn/T/ipykernel_26624/780535670.py", line 1, in <module>
    from keras.models import Sequential
  File "/Users/zhongyihao/anaconda3/lib/python3.10/site-packages/keras/__init__.py", line 3, in <module>
    from keras import __internal__
  File "/Users/zhongyihao/anaconda3/lib/python3.10/site-packages/keras/__internal__/__init__.py", line 6, in <module>
    from keras.__internal__ import models
  File "/Users/zhongyihao/anaconda3/lib/python3.10/site-packages/keras/__internal__/models/__init__.py", line 3, in <module>
    from keras.src.models.cloning import clone_and_build_model
  File "/Users/zhongyihao/anaconda3/lib/python3.10/site-packages/keras/src/__init__.py", line 21, in <module>
    from keras.src import applications
  F

### 3.2 






### 3.3. 



### 3.4. Discussion



## Problem 4 - 

### 4.1. Plot

#### 4.1. Written Answer



### 4.2. 



### 4.3. 

### 4.4. 

